inject/src/main/java/io/micronaut/context/AbstractBeanDefinition2.java

```diff
@@ -0,0 +1,1846 @@
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.AnnotationUtil;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.annotation.UsedByGeneratedCode;
+import io.micronaut.core.bind.annotation.Bindable;
+import io.micronaut.core.convert.ArgumentConversionContext;
+import io.micronaut.core.convert.ConversionContext;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.naming.Named;
+import io.micronaut.core.reflect.ClassUtils;
+import io.micronaut.core.reflect.ReflectionUtils;
+import io.micronaut.core.type.Argument;
+import io.micronaut.core.type.DefaultArgument;
+import io.micronaut.core.util.CollectionUtils;
+import io.micronaut.core.value.PropertyResolver;
+import io.micronaut.inject.BeanDefinition;
+import io.micronaut.inject.ConstructorInjectionPoint;
+import io.micronaut.inject.ExecutableMethod;
+import io.micronaut.inject.ExecutableMethodsDefinition;
+import io.micronaut.inject.FieldInjectionPoint;
+import io.micronaut.inject.MethodInjectionPoint;
+import io.micronaut.inject.ValidatedBeanDefinition;
+import io.micronaut.inject.annotation.AbstractEnvironmentAnnotationMetadata;
+import io.micronaut.inject.qualifiers.InterceptorBindingQualifier;
+import io.micronaut.inject.qualifiers.Qualifiers;
+import io.micronaut.inject.qualifiers.TypeAnnotationQualifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+/**
+ * <p>Default implementation of the {@link BeanDefinition} interface. This class is generally not used directly in user
+ * code.
+ * Instead a build time tool does analysis of source code and dynamically produces subclasses of this class containing
+ * information about the available injection points for a given class.</p>
+ * <p>
+ * <p>For technical reasons the class has to be marked as public, but is regarded as internal and should be used by
+ * compiler tools and plugins (such as AST transformation frameworks)</p>
+ * <p>
+ * <p>The {@link io.micronaut.inject.writer.BeanDefinitionWriter} class can be used to produce bean definitions at
+ * compile or runtime</p>
+ *
+ * @param <T> The Bean definition type
+ * @author Graeme Rocher
+ * @author Denis Stepanov
+ * @see io.micronaut.inject.writer.BeanDefinitionWriter
+ * @since 3.0
+ */
+@Internal
+public class AbstractBeanDefinition2<T> extends AbstractBeanContextConditional implements BeanDefinition<T>, EnvironmentConfigurable {
```

Comment:
Let's name this `AbstractInitializableBeanDefinition` since it is more descriptive of what it does vs the previous one